jlm/llvm/opt/alias-analyses/AliasAnalysis.hpp

```diff
@@ -0,0 +1,343 @@
+ */
+class LocalAliasAnalysis final : public AliasAnalysis
+{
+  // When doing origin tracing, give up if the trace set grows larger than this
+  static constexpr size_t MaxTraceCollectionSize = 1000;
+
+public:
+  LocalAliasAnalysis();
+  ~LocalAliasAnalysis() override;
+
+  std::string
+  ToString() const override;
+
+  AliasQueryResponse
+  Query(const rvsdg::Output & p1, size_t s1, const rvsdg::Output & p2, size_t s2) override;
+
+private:
+  struct TracedPointerOrigin;
+  struct TraceCollection;
+
+  /**
+   * Calculates the byte offset applied by the GEP, if the offset is static.
+   * The offset is the number of bytes needed to satisfy
+   *   output ptr = input ptr + offset in bytes
+   *
+   * @param gepNode the node representing the GEP operation
+   * @return the offset applied by the GEP, if it is possible to determine at compile time
+   */
+  [[nodiscard]] static std::optional<int64_t>
+  CalculateGepOffset(const rvsdg::SimpleNode & gepNode);
+
+  /**
+   * Returns the result of tracing the origin of p as far as possible, without introducing any
+   * uncertainty. By not introducing uncertainty, the result may be used to determine MustAlias
+   * relations. p is normalized, and traced through GEPs with compile time known offsets, as far as
+   * possible.
+   *
+   * @param p the pointer value to trace
+   * @return the TracedPointer for p, which is guaranteed to have a defined offset
+   */
+  [[nodiscard]] static TracedPointerOrigin
+  TracePointerOriginPrecise(const rvsdg::Output & p);
+
+  /**
+   * Given two pointers with the same base pointer
+   *  p1 = base + offset1
+   *  p2 = base + offset2
+   * Answers if the regions [p1, p1 + s1) and [p2, p2 + s2) may alias.
+   * If p1 == p2, MustAlias is returned.
+   * @param offset1 the offset of the first pointer. If it is nullopt, the offset is unknown.
+   * @param s1 the size of the first memory region
+   * @param offset2 the offset of the second pointer. If it is nullopt, the offset is unknown.
+   * @param s2 the size of the second memory region
+   */
+  [[nodiscard]] static AliasQueryResponse
+  QueryOffsets(
+      std::optional<int64_t> offset1,
+      size_t s1,
+      std::optional<int64_t> offset2,
+      size_t s2);
+
+  /**
+   * Traces to find all possible origins of the given pointer.
+   * Traces through GEP operations, including those with offsets that are not known at compile time.
+   * Also traces through gamma and theta nodes, building a set of multiple possibilities.
+   * Tracing stops at "top origins", for example an ALLOCA, a LOAD, or the return value of a CALL.
+   *
+   * @param p the pointer to trace from
+   * @param traceCollection the collection of trace points being created
+   * @return false if the trace collection reached its maximum allowed size, and tracing aborted
+   */
+  [[nodiscard]] static bool
+  TraceAllPointerOrigins(TracedPointerOrigin p, TraceCollection & traceCollection);
```

Comment:
Could this interface be simplified a bit to: `std::optional<TraceCollection> TraceAllPointerOrigins(TracedPointerOrigin p)`. In other words, we only get a TraceCollection if we did not reac the maximum allowed size?

Reply:
Nah, the function is recursive....not possible. Ignore this comment.